Implement recursive fmt::Debug for patrica_trie::TrieDB to print Node trees

gnunicorn opened the pull request:
fixes #8184

gnunicorn commented:
ah. I didn't see this part. I had run `cargo fmt` in between and it had changed... _a lot_ . I can revert that part, too, if you want  @dvdplm . I don't really have a preference...

debris commented on util/patricia_trie/src/node.rs:
```diff
@@ -107,13 +108,43 @@ impl<'a> Node<'a> {
+			Node::Branch(ref nodes, ref value) => {
+				let nodes_fmt = f.debug_list().entries(nodes).finish()?;
+				f.debug_struct("Node::Branch")
+					.field("nodes", &nodes_fmt)
```
I think this is invalid, cause `nodes_fmt` is `()`.? [docs](https://doc.rust-lang.org/std/fmt/struct.DebugList.html#method.finish)

debris commented on util/patricia_trie/src/node.rs:
```diff
@@ -107,13 +108,43 @@ impl<'a> Node<'a> {
+				let item = item.fmt(f)?;
+				f.debug_struct("Node::Extension")
+					.field("slice", &slice)
+					.field("item", &item)
```
I think this is invalid, cause `item` is `()`.

debris commented on util/patricia_trie/src/node.rs:
```diff
@@ -107,13 +108,43 @@ impl<'a> Node<'a> {
+		match &self {
+			Node::Leaf(slice, value) => f.debug_struct("Node::Leaf")
+				.field("slice", &slice)
+				.field("value", &value.pretty())
```
is it possible to avoid calling `.pretty()`? :)

gnunicorn replied:
Yes, but that gives a different value: `00·01` (on pretty) vs `[0, 1]`. I am fine with the latter, but just to make sure that is okay.

gnunicorn commented:
@debris just come to realize that you are right, for once because of the reasons you describe (formatter isn't really meant to be used recursively and the way I thought it worked was ... unsequential), but more importantly: the errors that were previously possible in the formatter came from the lookup. My current approach isn't actually doing the look up but just prints the addresses. Thus we are not seeing any tree actually.

Let me get back to this.

debris commented on util/patricia_trie/src/node.rs:
```diff
@@ -177,3 +171,142 @@ impl<'a> From<Node<'a>> for OwnedNode {
+		let empty = Node::Empty;
+
+		assert_eq!(format!("{:?}", empty), "Node::Empty");
+        assert_eq!(format!("{:#?}", empty), "Node::Empty");
```
bad indentation ;)

gnunicorn commented:
@debris yeah. I tried to push it into Node first (handing over a closure) and that might have worked but wouldn't actually take into account the right ident.

So, I moved the entire implementation into its own `TrieAwareDebugNode` that handles lookup and all nicely and added a test on the triedb-side of things.

But as I had to remove all I did before, I decided to rebase it entirely. I hope that is okay.

gnunicorn commented on util/patricia_trie/src/triedb.rs:
```diff
@@ -174,15 +119,55 @@ impl<'db> Trie for TrieDB<'db> {
+						.field("item", &TrieAwareDebugNode{trie: self.trie, key: item})
+					.finish(),
+				Ok(Node::Branch(ref nodes, ref value)) => f.debug_struct("Node::Branch")
+						.field("nodes", &nodes)
```
hold on a second, this still doesn't convert the entries.
